### PR TITLE
13.0 mig product variant sale price

### PR DESCRIPTION
**Pull request by jlaloux**

### CODE REVIEW

**Review by Jerther — COMMENTED** [review]
Hi @jlaloux 
Could you explain why in https://github.com/OCA/product-variant/pull/177/commits/5985cd9c2211a5eda93f30ed534d1fed0bb1f671 all single quotes had to be replaced by double quotes?

Also these small travis messages:

```
product_variant_sale_price/__manifest__.py:14:34: E231 missing whitespace after ','
product_variant_sale_price/__manifest__.py:15:39: E231 missing whitespace after ','
product_variant_sale_price/tests/test_product_product.py:15:69: E231 missing whitespace after ','
product_variant_sale_price/tests/test_product_product.py:18:63: E231 missing whitespace after ','
product_variant_sale_price/tests/test_product_product.py:21:62: E231 missing whitespace after ','
product_variant_sale_price/tests/test_product_product.py:58:30: E231 missing whitespace after ','
```

Still, nice work.

**Comment by jlaloux** [comment]
Hi @Jerther ,
I simply followed the steps explained here: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0

And that's the result after runing "pre-commit"

Regards,

Julien

**Review by Jerther — APPROVED** [review]
All good to me!

**Comment by Rad0van** [comment]
@jlaloux the way I understand OCA process is this won't be merged unless all the checks pass. So far the travis fails all tests. So you first need to fix the flake8 errors (i.e. `"depends": ["account", "sale",],` this should become `"depends": ["account", "sale"],`. Afterwards travis will attempt to run tests to see if everything works as it should. And currently the tests end up with error on `AttributeError: 'product.product' object has no attribute 'attribute_value_ids'`. That means there's something fundamentally wrong with the code (if it was problem with functionality the test would fail instead of end up with error). After those tests pass you can expect OK reviews which can lead to module being merged.
Will you please do those? If not I can do them as well because I need the module on 13.0. Thank you.

File: product_variant_sale_price/tests/test_product_product.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                            ],
+                        },
+                    )
+                ],
+                "uom_id": self.uom_unit.id,
+            }
+        )
+
+        self.product_blue = self.product_template.product_variant_ids.filtered(
+            lambda x: x.attribute_value_ids == self.att_color_blue
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:48]:
```suggestion
            lambda x: x.product_template_attribute_value_ids.product_attribute_value_id == self.att_color_blue
```

File: product_variant_sale_price/tests/test_product_product.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        )
+        self.product_red = self.product_template.product_variant_ids.filtered(
+            lambda x: x.attribute_value_ids == self.att_color_red
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:51]:
```suggestion
            lambda x: x.product_template_attribute_value_ids.product_attribute_value_id == self.att_color_red
```

File: product_variant_sale_price/tests/test_product_product.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            {"fix_price": 0.0,}
+        )
+        # Take account price extra
+        self.env["product.template.attribute.value"].create(
+            {
+                "product_tmpl_id": self.product_template.id,
+                "product_attribute_value_id": self.att_color_blue.id,
+                "price_extra": 100.00,
+            }
+        )
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:67]:
```suggestion
        self.product_blue.product_template_attribute_value_ids.write({"price_extra": 100.00})
```

File: product_variant_sale_price/tests/test_product_product.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        self.env["product.template.attribute.value"].create(
+            {
+                "product_tmpl_id": self.product_template.id,
+                "product_attribute_value_id": self.att_color_red.id,
+                "price_extra": 200.00,
+            }
+        )
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:74]:
```suggestion
        self.product_red.product_template_attribute_value_ids.write({"price_extra": 200.00})
```

File: product_variant_sale_price/tests/test_product_product.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+        set_sale_price_on_variant(self.cr, None, self.product_template.id)
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:76]:
```suggestion
        # Flush the records to DB before direct SQL
        self.product_blue.flush()
        self.product_red.flush()
        set_sale_price_on_variant(self.cr, None, self.product_template.id)
```

File: product_variant_sale_price/tests/test_product_product.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            }
+        )
+        self.env["product.template.attribute.value"].create(
+            {
+                "product_tmpl_id": self.product_template.id,
+                "product_attribute_value_id": self.att_color_red.id,
+                "price_extra": 200.00,
+            }
+        )
+
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:75]:
```suggestion
        # Flush the records to DB before direct SQL
        self.product_blue.flush()
        self.product_red.flush()

```

File: product_variant_sale_price/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+{
+    "name": "Product Variant Sale Price",
+    "summary": "Allows to write fixed prices in product variants",
+    "version": "13.0.1.0.0",
+    "category": "Product Management",
+    "website": "https://odoo-community.org/",
+    "author": "Tecnativa, " "Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "installable": True,
+    "depends": ["account", "sale",],
```
Comment (Rad0van) [inline product_variant_sale_price/__manifest__.py:14]:
```suggestion
    "depends": ["account", "sale"],
```

File: product_variant_sale_price/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    "data": ["views/product_views.xml",],
```
Comment (Rad0van) [inline product_variant_sale_price/__manifest__.py:15]:
```suggestion
    "data": ["views/product_views.xml"],
```

File: product_variant_sale_price/hooks.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+def set_sale_price_on_variant(cr, registry, template_id=None):
+    sql = """
+        UPDATE product_product pp
+        SET fix_price = pt.list_price + (
+            SELECT COALESCE(SUM(ptav.price_extra), 0)
+            FROM product_variant_combination pav_pp_rel
+            LEFT JOIN product_template_attribute_value ptav ON
+                ptav.product_attribute_value_id =
+                pav_pp_rel.product_template_attribute_value_id
+            WHERE pav_pp_rel.product_product_id = pp.id
```
Comment (Rad0van) [inline product_variant_sale_price/hooks.py:15]:
```suggestion
            FROM product_variant_combination pvc
            LEFT JOIN product_template_attribute_value ptav ON
                ptav.product_attribute_value_id =
                pvc.product_template_attribute_value_id
            WHERE pvc.product_product_id = pp.id
```

File: product_variant_sale_price/tests/test_product_product.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+class TestProductVariantPrice(TransactionCase):
+    def setUp(self):
+        super(TestProductVariantPrice, self).setUp()
+        self.template = self.env["product.template"]
+        self.product_product = self.env["product.product"]
+        self.attribute = self.env["product.attribute"]
+        self.attribute_value = self.env["product.attribute.value"]
+
+        self.att_color = self.attribute.create({"name": "color_test",})
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:15]:
```suggestion
        self.product_product = self.env["product.product"]
        self.attribute = self.env["product.attribute"]
        self.attribute_value = self.env["product.attribute.value"]

        self.att_color = self.attribute.create({"name": "color_test"})
```

File: product_variant_sale_price/tests/test_product_product.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+        self.att_color_blue = self.attribute_value.create(
+            {"name": "Blue", "attribute_id": self.att_color.id,}
+        )
+        self.att_color_red = self.attribute_value.create(
+            {"name": "Red", "attribute_id": self.att_color.id,}
+        )
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:22]:
```suggestion
        self.att_color = self.attribute.create({"name": "color_test"})

        self.att_color_blue = self.attribute_value.create(
            {"name": "Blue", "attribute_id": self.att_color.id}
        )
        self.att_color_red = self.attribute_value.create(
            {"name": "Red", "attribute_id": self.att_color.id}
        )
```

**Comment by Rad0van** [comment]
@jlaloux I did the migration myself too with almost the same results. I provided few suggestions that should fix the travis as well as functional tests. Let me know please if you'd be going on with migration. Thanx.

File: product_variant_sale_price/tests/test_product_product.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        self.product_red = self.product_template.product_variant_ids.filtered(
+            lambda x: x.attribute_value_ids == self.att_color_red
+        )
+
+    def test_post_init_hook(self):
+        from ..hooks import set_sale_price_on_variant
+
+        self.product_template.product_variant_ids.write(
+            {"fix_price": 0.0,}
```
Comment (Rad0van) [inline product_variant_sale_price/tests/test_product_product.py:58]:
```suggestion
            {"fix_price": 0.0}
```

**Comment by wpichler** [comment]
I'd like to migrate this to version 14. Is there still any activity here ? Which branch should i use as base for a migriation to v14 ?

**Comment by Rad0van** [comment]
If @jlaloux is no longer active on this I can redo the PR. What's your opinion @Jerther? Then @wpichler can base the work on this version.

**Comment by Jerther** [comment]
@Rad0van I'm already using this code into production, and I have been for a month now with no problem. As far as I'm concerned, the sooner this PR gets merged, the better. If you do make another PR with your revision applied, just mention me and I'll quickly do a review.

**Comment by Rad0van** [comment]
> @Rad0van I'm already using this code into production, and I have been for a month now with no problem. As far as I'm concerned, the sooner this PR gets merged, the better. If you do make another PR with your revision applied, just mention me and I'll quickly do a review.

Me too :-) I'll do it in few minutes.

**Comment by wpichler** [comment]
I've tried the current v13 branch against v14 - it works without any changes needed...

**Comment by Jerther** [comment]
I think this should be closed since this module appears to have been migrated to 13.0 and 14.0.

**Pull request closed without merging**